src/libsyntax/parse/parser/item.rs

```diff
@@ -919,34 +915,53 @@ impl<'a> Parser<'a> {
     }
 
     /// Parses `trait Foo { ... }` or `trait Foo = Bar;`.
-    fn parse_item_trait(&mut self, is_auto: IsAuto, unsafety: Unsafety) -> PResult<'a, ItemInfo> {
+    fn parse_item_trait(
+        &mut self,
+        lo: Span,
+        is_auto: IsAuto,
+        unsafety: Unsafety
+    ) -> PResult<'a, ItemInfo> {
         let ident = self.parse_ident()?;
         let mut tps = self.parse_generics()?;
 
         // Parse optional colon and supertrait bounds.
-        let bounds = if self.eat(&token::Colon) {
+        let had_colon = self.eat(&token::Colon);
+        let span_at_colon = self.prev_span;
+        let bounds = if had_colon {
             self.parse_generic_bounds(Some(self.prev_span))?
         } else {
             Vec::new()
         };
 
+        let span_before_eq = self.prev_span;
         if self.eat(&token::Eq) {
             // It's a trait alias.
+            if span_at_colon != span_before_eq {
```

Comment:
```suggestion
            if had_colon {
```
?

Spans are unreliable and can all be dummy easily, parser shouldn't generally rely on them.

Also, if this is a syntactic error (as opposed to a semantic error), then empty bound list `trait Alias: /* nothing */ = Bounds;` should also be an error.

Reply:
Ah, wasn't sure about the approach I was using, so it's great you pointed it out, thanks. Will fix. :)